packages/pubsub/src/subscription.js

```diff
@@ -592,30 +365,249 @@ Subscription.prototype.decorateMessage_ = function(message) {
  *   var apiResponse = data[0];
  * });
  */
-Subscription.prototype.delete = function(callback) {
+Subscription.prototype.delete = function(gaxOpts, callback) {
   var self = this;
 
-  callback = callback || common.util.noop;
+  if (is.fn(gaxOpts)) {
+    callback = gaxOpts;
+    gaxOpts = {};
+  }
 
-  var protoOpts = {
-    service: 'Subscriber',
-    method: 'deleteSubscription'
+  var reqOpts = {
+    subscription: this.name
   };
 
+  this.request({
+    client: 'subscriberClient',
+    method: 'deleteSubscription',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, function(err, resp) {
+    if (!err) {
+      self.removeAllListeners('message');
+    }
+
+    callback(err, resp);
+  });
+};
+
+/**
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this
+ *     request.
+ * @param {object} callback.apiResponse - Raw API response.
+ */
+Subscription.prototype.getMetadata = function(gaxOpts, callback) {
+  if (is.fn(gaxOpts)) {
+    callback = gaxOpts;
+    gaxOpts = {};
+  }
+
   var reqOpts = {
     subscription: this.name
   };
 
-  this.parent.request(protoOpts, reqOpts, function(err, resp) {
+  this.request({
+    client: 'subscriberClient',
+    method: 'getSubscription',
+    reqOpts: reqOpts,
+    gaxOpts: gaxOpts
+  }, callback);
+};
+
+/**
+ * Begin listening for events on the subscription. This method keeps track of
+ * how many message listeners are assigned, and then removed, making sure
+ * polling is handled automatically.
+ *
+ * As long as there is one active message listener, the connection is open. As
+ * soon as there are no more message listeners, the connection is closed.
+ *
+ * @private
+ *
+ * @example
+ * subscription.listenForEvents_();
+ */
+Subscription.prototype.listenForEvents_ = function() {
+  var self = this;
+
+  this.on('newListener', function(event) {
+    if (event === 'message') {
+      self.messageListeners++;
+
+      if (!self.connection) {
+        self.openConnection_();
+      }
+    }
+  });
+
+  this.on('removeListener', function(event) {
+    if (event === 'message' && --self.messageListeners === 0) {
+      self.closeConnection_();
+    }
+  });
+};
+
+/**
+ *
+ */
+Subscription.prototype.createMessage_ = function(data) {
+  var message = new Message(self, data);
+  this.messages_.add(message);
```

Comment:
My original point was about this.messages_. Why hold on to Message objects in the Subscription object? Seems like a memory leak to me. Do we ever need to use the Message payload again in the Subscription object? If not, then we shouldn't refer to it. The garbage collector can only do its job if code drops references to things it no longer needs.